### PR TITLE
Cloud alerts

### PR DESCRIPTION
**Pull request by jwitz**
Closes https://app.zenhub.com/workspaces/docs-5fb535258b4e61000eac76cf/issues/astronomer/issues/2283. Technical reviewers, please help me answer the following questions: 

- Could any of the alerts benefit from more detailed information on the conditions that fire the alerts? For instance, is it helpful to define the 'acceptable rate' for a heartbeat in the AirflowSchedulerUnhealthy alert? What is the actual storage limit for temporary data? 
- Are the follow-up steps for each alert accurate? I'm particularly worried about which resource allocations I tell people to adjust in the follow-ups that mention Airflow 2.0.
- Should we tell customers to mention anything specific when following the 'Contact your Astronomer representative' follow-up?

### CODE REVIEW

**Comment by jwitz** [comment]
Adding @wolfier and @krisdock for review per @paolaperaza's recommendation. I'd really appreciate your review on the follow-up actions, which are new for this draft and might save us some support ticket effort in the future.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 1]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
-| `AirflowFailureRate` | Airflow tasks are failing at a higher rate than normal. |
-| `AirflowSchedulerUnhealthy` | Airflow scheduler is unhealthy: heartbeat has dropped below the acceptable rate. |
-| `AirflowPodQuota` | Deployment is near its pod quota: it's been using over 95% of it's pod quota for over 10 minutes. |
-| `AirflowCPUQuota` | Deployment is near its CPU quota: it's been using over 95% of it's CPU quota for over 10 minutes. |
-| `AirflowMemoryQuota` | Deployment is near its memory quota: it's been using over 95% of it's memory quota for over 10 minutes. |
-
+## Deployment Alerts on Astronomer
 
-### Anatomy of a Deployment Alert
+Refer to the following chart for more information on each deployment-level alert you might receive from Astronomer, as well as follow-up steps for improving the health of your affected Deployments.
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:149]:
```suggestion
Refer to the chart below for more information on each Deployment-level alert you might receive from Astronomer, as well as corresponding remediation steps we recommend for your affected Deployment.
```

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 2]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
-        tier: airflow
-        component: scheduler
-        deployment: "{{ $labels.deployment }}"
-      annotations:
-        summary: "{{ $labels.deployment }} scheduler is unhealthy"
-        description: "The {{ $labels.deployment }} scheduler's heartbeat has dropped below the acceptable rate."
-```
+| Alert | Description | Follow-Up |
+| ------------- | ------------- | ------------- |
+| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Contact your Astronomer representative. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:153]:
```suggestion
| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

Astronomer Cloud users/customers don't have a "representative" since our CS team doesn't have enough dedicated resources to reach all of them directly (and they likely won't be able to respond in an actionable way to notice of these alerts either), so I think our best bet is to send these folks to our support team.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 3]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Ensure that you are continually removing unused temporary data in your Airflow tasks. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:154]:
```suggestion
| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Make sure to continually remove unused temporary data in your Airflow tasks. |
```

Or even: 

`Remove unused temporary data in your Airflow tasks.`

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 4]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's allocated Webserver and Scheduler Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
+| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Contact your Astronomer representative. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:156]:
```suggestion
| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 5]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated Server Resources in the Astronomer UI or redesign your DAGs so that less tasks are concurrently pending. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:157]: Yikes really not sure what the recommended action item is here, so double check with @wolfier or @krisdock here - these guidelines sound like they're more telling you what's happening and what "should" happen but not so much how to get there.

When we mention `Server resources` though - is it Webserver, Scheduler, or Worker AU's they have to increase?

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 6]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `ContainerMemoryNearTheLimitInDeployment` | The Airflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 10 minutes. | Either increase the Deployment's allocated Webserver Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:158]: Do we know what the memory quota is? The Airflow Webserver just parses DAG files and powers the Airflow UI so I'm not immediately seeing a connection - maybe just double check? Is it Webserver, Scheduler or Worker resources that would help? Or some combination?

**Review by paolaperaza — CHANGES_REQUESTED** [review]
Thanks for this @jwitz! I just left a few comments here. I'm mostly not confident in approving the recommended action items since I'm not as familiar with all of these use cases - will lean on you @krisdock @wolfier.

The more specific we can be of what is happening, the better - how much Memory/CPU? What's the max ephemeral storage?

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 7]
```diff
@@ -1,4 +1,4 @@
----
+Deployment-level---
```
[the remark this answers is not in the document]
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:1]: Done!

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 1 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:149]: Made a different suggestion on this, since I realized this my original wording is grammatically inconsistent (the follow-up step _is_ more information, so we are not showing that "as well as" more information). Also tied to make it more active.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 5 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:157]: Whoops - I meant to write Webserver Resources there. And yeah, I agree we shouldn't move forward with this until Allan or Kris get eyes on it.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 8]
```diff
@@ -134,39 +131,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
-        deployment: "{{ $labels.deployment }}"
-      annotations:
-        summary: "{{ $labels.deployment }} scheduler is unhealthy"
-        description: "The {{ $labels.deployment }} scheduler's heartbeat has dropped below the acceptable rate."
-```
+| Alert | Description | Follow-Up |
+| ------------- | ------------- | ------------- |
+| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Contact your Astronomer representative. |
+| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Ensure that you are continually removing unused temporary data in your Airflow tasks. |
+| `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's allocated Webserver and Scheduler Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
```
[the remark this answers is not in the document]
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:155]: Assuming this was for the `AiflowPodQuota`. Should this be the only follow-up to recommend the Extra Capacity slider, rather than the allocated resources sliders?

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 9]
```diff
@@ -134,39 +131,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Contact your Astronomer representative. |
+| `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated Server Resources in the Astronomer UI or redesign your DAGs so that less tasks are concurrently pending. |
+| `ContainerMemoryNearTheLimitInDeployment` | The Airflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 10 minutes. | Either increase the Deployment's allocated Webserver Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
```
[the remark this answers is not in the document]
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:158]: Sounds good! Updated

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (krisdock) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md]: @jwitz correct. the resource sliders are specific to cpu/mem on the specific related components (webserver, scheduler, workers). the only thing that affects the pod quota is the extra capacity slider.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 6 above)
Comment (wolfier) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md]: @jwitz, you should check the actual [prometheus alerts](https://prometheus.gcp0001.us-east4.astronomer.io/rules) to verify the definition. For example `ContainerMemoryNearTheLimitInDeployment` is the collection of all the pods with some exceptions.

```
 (container_memory_working_set_bytes{container=~".*(scheduler|scheduler-gc|webserver|worker|statsd|pgbouncer|metrics-exporter|redis|flower)",container_name!~"POD|istio-proxy"} / container_spec_memory_limit_bytes{container=~".*(scheduler|scheduler-gc|webserver|worker|statsd|pgbouncer|metrics-exporter|redis|flower)",container_name!~"POD|istio-proxy"}) * 100 > 95 < +Inf
```

I would ask someone on infra to help you understand what those metrics mean.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 5 above)
Comment (wolfier) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md]: Server resources is most likely referring to the scheduler since the scheduler is what pushes the tasks through its lifecycle. However, since tasks are going into the scheduled/queued state that means the scheduler is functioning correctly most likely. The executor could be at fault for not sending the tasks to the workers.

The more likely scenario is that the worker does not have the ability to consume the tasks from the queue. Then the solution becomes increasing worker productivity* for the Executor, which are generally KubernetesExecutor or CeleryExecutor.

Increasing worker productivity is a subject for another day ;p 

Basically tuning the worker/scheduler would resolve this alert and I don't think it's as simple as bumping resources.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 6 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md]: Sounds good! Updated all of the inaccurate thresholds in the latest pull

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 10]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
-        deployment: "{{ $labels.deployment }}"
-      annotations:
-        summary: "{{ $labels.deployment }} scheduler is unhealthy"
-        description: "The {{ $labels.deployment }} scheduler's heartbeat has dropped below the acceptable rate."
-```
+| Alert | Description | Follow-Up |
+| ------------- | ------------- | ------------- |
+| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Reach out to [Astronomer Support](https://support.astronomer.io). |
+| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Make sure to continually remove unused temporary data in your Airflow tasks. |
+| `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's Extra Capacity in the Astronomer UI or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:155]: Typo on `over 95% of it's pod quota` --> `over 95% of its pod quota`

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 11]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:156]:
```suggestion
| `AirflowSchedulerUnhealthy` | The Airflow Scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

I think we typically capitalize Scheduler + Webserver - maybe something to standardize? Do we know what the acceptable rate of the heartbeat is?

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 12]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated resources in the Astronomer UI or redesign your DAGs so that fewer tasks are concurrently pending. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:157]: - Should this be "If your tasks _aren't_ running as expected, increase...."?
- What resources should they modify?
- Instead of "Redesign your DAGs" I'd say -  "Lower Airflow Concurrency or Parallelism so that fewer tasks are concurrently pending."

Comment (paolaperaza) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:157]: Could maybe even link out to: https://www.astronomer.io/guides/airflow-scaling-workers

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md [hunk 13]
```diff
@@ -136,39 +136,24 @@ The code in your DAG might look something like this ([source](https://github.com
...
+| `ContainerMemoryNearTheLimitInDeployment` | A container in your Aiirflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 60 minutes. | Either increase the Deployment's allocated resources in the Astronomer UI or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
```
Comment (paolaperaza) [inline cloud/stable/04_customize-airflow/04_airflow-alerts.md:158]: typo on "it's" v" its"

Comment (paolaperaza) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:158]: (on `ContainerMemoryNearTheLimitInDeployment`)

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 12 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:157]: I think they should do these things if their tasks _are_ running as expected, eg. their DAGs are properly creating and running tasks, but there still isn't enough resources to handle the number of pending tasks they have. 

Based on the article you linked out, I assume that the allocated resource they need to update will vary based on how they decide to increase concurrency/parallelism. In that case, I thin kit makes sense to link out to the guide for more information.

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 13 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:158]: Fixed!

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 11 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:156]: Good to know, I'll add that to the style guide.

It looks to be "no heartbeat for more than one minute" based on this line in the Prometheus code: `rate(airflow_scheduler_heartbeat{type="counter"}[1m]) == 0`
I'll add that in!

File: cloud/stable/04_customize-airflow/04_airflow-alerts.md (hunk 10 above)
Comment (jwitz) [reply cloud/stable/04_customize-airflow/04_airflow-alerts.md:155]: bleh, fixed

**Review by paolaperaza — APPROVED** [review]
Awesome @jwitz - this final round looks good to me!
